### PR TITLE
MM-26696: Add v2 APIs to circleCI package

### PR DESCRIPTION
**Pull request by agnivade**
#### Ticket Link

https://mattermost.atlassian.net/browse/MM-26696

### CODE REVIEW

**Review by metanerd — APPROVED** [review]
Thank you, that bothered me for quite some time already!

File: circleci.go [hunk 1]
```diff
@@ -846,6 +847,105 @@ func (c *Client) AddHerokuKeyWithContext(ctx context.Context, key string) error
...
+// TriggerPipeline calls TriggerPipelineWithContext with context.Background.
+func (c *Client) TriggerPipeline(vcsType VcsType, account, repo, branch, tag string, params map[string]interface{}) (*Pipeline, error) {
+	return c.TriggerPipelineWithContext(context.Background(), vcsType, account, repo, branch, tag, params)
+}
+
+// TriggerPipeline triggers a new pipeline for the given project for the given branch or tag.
+// Note that branch and tag are mutually exclusive. It is not exactly specified in the documentation
+// which takes precendence when both are specified.
+// https://circleci.com/docs/api/v2/?shell#trigger-a-new-pipeline
+// Note that this is only available as a v2 API.
```
Comment (ethervoid) [inline circleci.go:1013]: I'm not convinced about mixing v1 and v2 in the same client.

Ideally, we should have different clients struct or interfaces for v1 and v2 so we can create a different client based on the version but that would be a lot of work and I think that adding a `version` property could work.

The version property would:

- Set the version in the path by default
- Filter methods returning errors if the method is not available in the client version.

That way, given that the final user is going to use this library, we can protect them and return a proper error to them.

File: circleci.go [hunk 2]
```diff
@@ -846,6 +847,105 @@ func (c *Client) AddHerokuKeyWithContext(ctx context.Context, key string) error
 	return c.request(ctx, "POST", "/user/heroku-key", nil, nil, body)
 }
 
+// Pipeline describes a pipeline object.
+type Pipeline struct {
+	// The unique ID of the pipeline.
+	ID string `json:"id"`
```
Comment (ethervoid) [inline circleci.go:995]: What about including the `errors` and `project_slug` properties?

File: circleci.go [hunk 3]
```diff
@@ -846,6 +847,105 @@ func (c *Client) AddHerokuKeyWithContext(ctx context.Context, key string) error
...
+		Branch     string                 `json:"branch"`
+		Tag        string                 `json:"tag"`
+		Parameters map[string]interface{} `json:"parameters"`
+	}{
+		Branch:     branch,
+		Tag:        tag,
+		Parameters: params,
+	}
+
+	err := c.request(ctx, http.MethodPost, fmt.Sprintf("project/%s/%s/%s/pipeline", vcsType, account, repo), &p, nil, body)
```
Comment (ethervoid) [inline circleci.go:1029]: I'm seeing the [trigger new pipeline](https://circleci.com/docs/api/v2/#trigger-a-new-pipeline) method and that method's path doesn't correspond with the one we have here. In the docs is `/project/{project-slug}/pipeline.

Also, it doesn't return anything. It only returns a 201 status

**Review by ethervoid — COMMENTED** [review]
Some comments to discuss about it

File: circleci.go (hunk 2 above)
Comment (agnivade) [reply circleci.go:995]: Not sure what you mean here. This is the response that we get from the endpoint. Can you elaborate how do you want to integrate errors/project_slug into this?

File: circleci.go (hunk 3 above)
Comment (agnivade) [reply circleci.go:1029]: project-slug actually means `vcsType/account/repo` :) See the description of the field.

Can you clarify where you are seeing it doesn't return anything? The response schema clearly shows what it returns.

File: circleci.go (hunk 1 above)
Comment (agnivade) [reply circleci.go:1013]: Yeah I think I agree with your thoughts.

To confirm, you are proposing to add a `version` field to the struct that a user will set during initialization. Perhaps a new constructor is better here which sets the URL from the version?

And then add checks in every method which confirms which version the client is using and returns an appropriate error if there's a mismatch.

Does that match with what you were thinking?

Comment (ethervoid) [reply circleci.go:1013]: Yes that is what I'm thinking :)

File: circleci.go (hunk 2 above)
Comment (ethervoid) [reply circleci.go:995]: Sorry I was going to remove this comment and I miss it hahah.

File: circleci.go (hunk 3 above)
Comment (ethervoid) [reply circleci.go:1029]: Oh ok, I missread it then :facepalm:

**Review by ethervoid — APPROVED** [review]
Really nice one, love it. My only comments were fixed with the “final touches commit” 🚀

**Pull request merged**